gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/store/ValidUpdateOperation.java

```diff
@@ -0,0 +1,143 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+package com.pivotal.gemfirexd.internal.engine.store;
+
+import com.gemstone.gemfire.LogWriter;
+import com.gemstone.gemfire.cache.Region;
+import com.pivotal.gemfirexd.internal.catalog.types.RoutineAliasInfo;
+import com.pivotal.gemfirexd.internal.engine.Misc;
+import com.pivotal.gemfirexd.internal.engine.distributed.utils.GemFireXDUtils;
+import com.pivotal.gemfirexd.internal.iapi.error.StandardException;
+import com.pivotal.gemfirexd.internal.iapi.services.context.ContextService;
+import com.pivotal.gemfirexd.internal.iapi.sql.Activation;
+import com.pivotal.gemfirexd.internal.iapi.sql.PreparedStatement;
+import com.pivotal.gemfirexd.internal.iapi.sql.ResultSet;
+import com.pivotal.gemfirexd.internal.iapi.sql.conn.LanguageConnectionContext;
+import com.pivotal.gemfirexd.internal.iapi.sql.conn.StatementContext;
+import com.pivotal.gemfirexd.internal.iapi.types.DataValueDescriptor;
+import com.pivotal.gemfirexd.internal.impl.jdbc.EmbedConnection;
+import com.pivotal.gemfirexd.internal.impl.sql.GenericParameterValueSet;
+import com.pivotal.gemfirexd.internal.impl.sql.GenericPreparedStatement;
+
+/**
+ * This class does the validation if an update statement is pk plus some conditions.
+ * Right now we only support AND conditions. In future might add more.
+ */
+public class ValidUpdateOperation {
+  public static boolean isValid(Region<?, ?> region, String predicateString,
+      DataValueDescriptor[] otherKeyValues) throws
+      StandardException {
+
+
+    ResultSet r = null;
+    EmbedConnection conn = null;
+    StatementContext statementContext = null;
+    boolean contextSet = false;
+    boolean popContext = false;
+    Throwable t = null;
+    // setup LCC and get the ResultSet
+    LanguageConnectionContext lcc = Misc.getLanguageConnectionContext();
+    try {
+
+      if (lcc == null) {
+        //a PK based insert is converted into
+        // region.put since it bypasses GemFireXD layer, the LCC can be null.
+        conn = GemFireXDUtils.getTSSConnection(true, true, false);
+        conn.getTR().setupContextStack();
+        lcc = conn.getLanguageConnectionContext();
+
+
+        // lcc can be null if the node has started to go down.
+        if (lcc == null) {
+          Misc.getGemFireCache().getCancelCriterion()
+              .checkCancelInProgress(null);
+        }
+      }
+
+      final GemFireContainer container = (GemFireContainer)region
+          .getUserAttribute();
+      String sql = "SELECT * FROM " + container.getQualifiedTableName() + " WHERE " +
+          predicateString;
+
+      PreparedStatement queryStatement = lcc.prepareInternalStatement(sql,
+          (short)0);
+
```

Comment:
This is totally in for trouble. We cannot execute another query within region.delta apply. We need to use ExpressionBuilder and compile and evaluate just the predicate string.